### PR TITLE
Generate proper bits of random value for LongDecimalType tests.

### PR DESCRIPTION
**Pull request by yiweiHeOSS**
This PR is to fix https://github.com/facebookincubator/velox/issues/6317. The explanation of the fix is in the issue.

### CODE REVIEW

File: velox/dwio/common/tests/utils/BatchMaker.cpp [hunk 1]
```diff
@@ -184,11 +184,19 @@ VectorPtr BatchMaker::createVector<TypeKind::HUGEINT>(
     MemoryPool& pool,
     std::mt19937& gen,
     std::function<bool(vector_size_t /*index*/)> isNullAt) {
+  int bitsToMove = 0;
+  // Generate proper bits of random value for LongDecimalType tests.
+  if (type->isLongDecimal()) {
+    auto [precision, scale] = getDecimalPrecisionScale(*type);
+    int bitsNeeded = ceil(log2(std::pow(10, precision))) + 1;
```
Comment (yingsu00) [inline velox/dwio/common/tests/utils/BatchMaker.cpp:191]: why + 1? 
also, strictly speaking, the maximum #digits `precision` can express is std::pow(10, precision) - 1.
SHould it be `ceil(log2(std::pow(10.0, precision) - 1));`?

File: velox/dwio/common/tests/utils/BatchMaker.cpp [hunk 2]
```diff
@@ -184,11 +184,19 @@ VectorPtr BatchMaker::createVector<TypeKind::HUGEINT>(
     MemoryPool& pool,
     std::mt19937& gen,
     std::function<bool(vector_size_t /*index*/)> isNullAt) {
+  int bitsToMove = 0;
+  // Generate proper bits of random value for LongDecimalType tests.
+  if (type->isLongDecimal()) {
```
[the remark this answers is not in the document]
Comment (yiweiHeOSS) [reply velox/dwio/common/tests/utils/BatchMaker.cpp:189]: The Shortdecimal will go to the BatchMaker::createVector<TypeKind::BIGINT>(). And it looks like its tests are working well.

File: velox/dwio/common/tests/utils/BatchMaker.cpp (hunk 1 above)
Comment (yiweiHeOSS) [reply velox/dwio/common/tests/utils/BatchMaker.cpp:191]: This is just a rough calculation. I don't know if there is a better way to do this calculation. But in the case it is DECIMAL(38), we want it to create data with 128 digits, right? In that case, `ceil(log2(std::pow(10, precision))) + 1;` is equal to 128.

Comment (yiweiHeOSS) [reply velox/dwio/common/tests/utils/BatchMaker.cpp:191]: The 1 extra digit (the most significant digit (MSD)) I think is for the +/- sign.

File: velox/dwio/common/tests/utils/BatchMaker.cpp [hunk 3]
```diff
@@ -184,11 +184,19 @@ VectorPtr BatchMaker::createVector<TypeKind::HUGEINT>(
...
+    auto [precision, scale] = getDecimalPrecisionScale(*type);
+    int bitsNeeded = ceil(log2(std::pow(10, precision))) + 1;
+    bitsToMove = 128 - bitsNeeded;
```
[the remark this answers is not in the document]
Comment (yiweiHeOSS) [reply velox/dwio/common/tests/utils/BatchMaker.cpp:192]: I just added it for better readability, but I can inline it if you like.

File: velox/dwio/common/tests/utils/BatchMaker.cpp [hunk 4]
```diff
@@ -184,11 +184,23 @@ VectorPtr BatchMaker::createVector<TypeKind::HUGEINT>(
...
     std::mt19937& gen,
     std::function<bool(vector_size_t /*index*/)> isNullAt) {
+  int bitsToMove = 0;
+  // Generate proper bits of random value for LongDecimalType tests.
+  if (type->isLongDecimal()) {
+    auto [precision, scale] = getDecimalPrecisionScale(*type);
+    int bitsNeeded = ceil(log2(std::pow(10, precision)));
+    // Round up if the bit number is not the multiples of 8 (1 byte).
+    if (bitsNeeded % 8 != 0) {
+      bitsNeeded = (bitsNeeded / 8 + 1) * 8;
```
Comment (yingsu00) [inline velox/dwio/common/tests/utils/BatchMaker.cpp:194]: I think it's just 
```
int bitsToMove = 128 - ceil(log2(std::pow(10, precision)) / 8) * 8;
```

Comment (yiweiHeOSS) [reply velox/dwio/common/tests/utils/BatchMaker.cpp:194]: Can I just keep it as it is if there is no major problem? I think it's more easy to understand and I already add it to another PR. Also, it looks like the `int bitsToMove = 128 - ceil(log2(std::pow(10, precision)) / 8) * 8;` is not right when Decimal(38).

**Review by yingsu00 — APPROVED** [review]

**Comment by yingsu00** [comment]
@Yuhta Can you please review this PR? Thanks!

**Comment by yiweiHeOSS** [comment]
Looks like this PR is already delivered to [facebookincubator:main](https://github.com/facebookincubator/velox/tree/main)

**Comment by Yuhta** [comment]
Merged in #6275

**Pull request closed without merging**